File: src/main/java/no/skatteetaten/aurora/databasehotel/dao/DataSourceUtils.java
package no.skatteetaten.aurora.databasehotel.dao;

import static com.google.common.base.Strings.nullToEmpty;

import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.zaxxer.hikari.HikariConfig;
import com.zaxxer.hikari.HikariDataSource;

public final class DataSourceUtils {

    private static final Logger LOGGER = LoggerFactory.getLogger(DataSourceUtils.class);

    private DataSourceUtils() {

    }

    public static HikariDataSource createDataSource(String jdbcUrl, String username, String password) {

        return createDataSource(jdbcUrl, username, password, 2);
    }

    public static HikariDataSource createDataSource(String jdbcUrl, String username, String password,
        int maximumPoolSize) {

        HikariConfig config = createConfig(jdbcUrl, username, password, maximumPoolSize);

        return createDataSource(config);
    }

    public static HikariDataSource createDataSource(HikariConfig config) {

        String maskedPassword = createPasswordHint(config.getPassword());
        LOGGER
            .debug("Creating datasource using jdbcUrl: \"{}\", username: \"{}\", password: \"{}\"", config.getJdbcUrl(),
                config.getUsername(), maskedPassword);

        return new HikariDataSource(config);
    }

    public static HikariConfig createConfig(String jdbcUrl, String username, String password, int maximumPoolSize) {

        HikariConfig config = new HikariConfig();
        config.setJdbcUrl(jdbcUrl);
        config.setUsername(username);
        config.setPassword(password);
        config.setMaximumPoolSize(maximumPoolSize);

        // Just some defaults. Has not been properly evaluated.
        config.addDataSourceProperty("cachePrepStmts", "true");
        config.addDataSourceProperty("prepStmtCacheSize", "250");
        config.addDataSourceProperty("prepStmtCacheSqlLimit", "2048");
        return config;
    }

    static String createPasswordHint(String password) {

        if (nullToEmpty(password).length() < 8) {
            return StringUtils.repeat("*", nullToEmpty(password).length());
        }
        return password.substring(0, 2) + StringUtils.repeat("*", password.length() - 4) + password
            .substring(password.length() - 2);
    }
}

File: src/main/kotlin/no/skatteetaten/aurora/databasehotel/dao/DataSourceUtils.kt
package no.skatteetaten.aurora.databasehotel.dao

import com.zaxxer.hikari.HikariConfig
import com.zaxxer.hikari.HikariDataSource
import mu.KotlinLogging
import org.apache.commons.lang3.StringUtils

private val logger = KotlinLogging.logger {}

private val LOGIN_TIMEOUT_SEC = 10

object DataSourceUtils {

    @JvmOverloads
    fun createDataSource(
        jdbcUrl: String,
        username: String,
        password: String,
        maximumPoolSize: Int = 2,
        loginTimeoutSec: Int = LOGIN_TIMEOUT_SEC
    ): HikariDataSource {

        val config = createConfig(jdbcUrl, username, password, maximumPoolSize)

        return createDataSource(config, loginTimeoutSec)
    }

    fun createDataSource(config: HikariConfig, loginTimeoutSec: Int = LOGIN_TIMEOUT_SEC): HikariDataSource {

        val maskedPassword = createPasswordHint(config.password)
        logger
            .debug(
                "Creating datasource using jdbcUrl: \"{}\", username: \"{}\", password: \"{}\"", config.jdbcUrl,
                config.username, maskedPassword
            )

        val hikariDataSource = HikariDataSource(config)
        // Setting this will prevent establishing connections from hanging forever in cases where the server is unable
        // to report back a connection error (for instance when under heavy load).
        hikariDataSource.loginTimeout = loginTimeoutSec
        return hikariDataSource
    }

    fun createConfig(jdbcUrl: String, username: String, password: String, maximumPoolSize: Int): HikariConfig {

        val config = HikariConfig()
        config.jdbcUrl = jdbcUrl
        config.username = username
        config.password = password
        config.maximumPoolSize = maximumPoolSize

        // Just some defaults. Has not been properly evaluated.
        config.addDataSourceProperty("cachePrepStmts", "true")
        config.addDataSourceProperty("prepStmtCacheSize", "250")
        config.addDataSourceProperty("prepStmtCacheSqlLimit", "2048")
        return config
    }

    fun createPasswordHint(password: String?): String {
        return if (password == null || password.length < 8) {
            StringUtils.repeat("*", password?.length ?: 0)
        } else {
            password.substring(0, 2) + StringUtils.repeat(
                "*",
                password.length - 4
            ) + password
                .substring(password.length - 2)
        }
    }
}
